File: solutions/haiku-1/haiku/python/anagrams.py
import sys

# Read all words
words = []
for line in sys.stdin:
    word = line.strip()
    if word:
        words.append(word)

# Group anagrams by their canonical form (sorted letters)
groups_dict = {}
for word in words:
    canonical = ''.join(sorted(word))
    if canonical not in groups_dict:
        groups_dict[canonical] = []
    groups_dict[canonical].append(word)

# Sort words within each group and collect all groups
groups = []
for word_list in groups_dict.values():
    word_list.sort()  # Sort alphabetically
    groups.append(word_list)

# Sort groups by their first word
groups.sort(key=lambda g: g[0])

# Output
for group in groups:
    print(' '.join(group))

File: solutions/haiku-1/haiku/perl/anagrams.pl
#!/usr/bin/perl
use strict;
use warnings;

my %groups;

# Read words and group by canonical form (sorted letters)
while (my $word = <STDIN>) {
    chomp $word;
    # Create canonical form by sorting letters
    my $canonical = join('', sort split //, $word);
    push @{$groups{$canonical}}, $word;
}

# Build output groups with sorted words
my @output;
for my $canonical (keys %groups) {
    my @sorted_words = sort @{$groups{$canonical}};
    push @output, \@sorted_words;
}

# Sort groups by their first word (alphabetically)
@output = sort { $a->[0] cmp $b->[0] } @output;

# Print each group
for my $group (@output) {
    print join(' ', @$group) . "\n";
}
